### ISSUE TITLE
performance degradation with gatoms in Pd 0.52-test2

### ISSUE THREAD

**Issue by Spacechild1** [pure-data/pure-data#1459]
The following patch contains a large amount of float gatoms:

[many-gatoms.zip](https://github.com/pure-data/pure-data/files/7559225/many-gatoms.zip)

On my system (Windows 7), the idle CPU load of Pd without DSP is ca. 0.25%

When I open this patch with Pd 0.51-4 and move the mouse, the CPU load rises up to a max. of 0.5%.

When I do the same with Pd 0.52-test1 or Pd 0.52-test2, the CPU load rises up to a max. of 5%!! On my 4-core laptop, this is 1/5 of a core just for moving the mouse!

Can someone confirm this?

**Comment by Lucarda** [comment]
`many-gatoms.pd`

Intel Core i5-520M
Windows 10
amd64
DSP On

|Pd 0.51-4 | Pd 0.52-test2|
|:----|:----|
|~2%|~20%|

**Comment by Ant1r** [comment]
Same here: 
Ubuntu 16.04 / AMD Phenom II forced to 2.2GHz

| Pd 0.51-4 | Pd 0.52-test2 |
|         ----    |     -------          |
| ~2%         | ~22%            |

**Comment by umlaeute** [comment]
just to clarify: this is about moving the mouse across the patch without any buttons pressed (that is: it's not about updating the gatoms, but their checking whether the mouse is over them)?

and consequently, the CPU usage is measured in the `pd`(-core) process rather than the `pd-gui`?

**Comment by Ant1r** [comment]
Yes, for my part I was just moving the mouse without any button pressed, and I measured the CPU load with the `Media/Load Meter`.

**Comment by Lucarda** [comment]
Same as @Ant1r , I measured it with Windows "task manager".

**Comment by Spacechild1** [comment]
@umlaeute yes and yes!

**Comment by Lucarda** [comment]
This has to be looked deeply but i found that commenting

https://github.com/pure-data/pure-data/blob/master/src/g_canvas.c#L499

improves back performance and gives clues also for #1410

**Comment by Lucarda** [comment]
This has come down substantially to normal figures


|Pd 0.51-4|	Pd 0.52-test4|
|:--|:--|
|~2%	|~3% |

**Comment by porres** [comment]
> This has come down substantially to normal figures

with what? latest merges?

**Comment by Lucarda** [comment]
Not sure exactly where. Now testing 621fb52 "**Pd 0.52-test4**"

The degradations above where tested under "**Pd 0.52-test2**"

**Comment by Spacechild1** [comment]
For me (Windows 7):

| Pd 0.51-4 | Pd 0.52-test4
| ---------- | --------------
| max. 0.45% | max. 4.5%

The idle CPU usage is 0.2%.

So it's not really fixed on my side...

**Comment by Lucarda** [comment]
All my tests were done with the Windows task manager on Win10.

**Comment by Spacechild1** [comment]
@Ant1r can you also test again?

**Comment by Lucarda** [comment]
Using Pd load meter I got:


|Pd 0.51-4|	Pd 0.52-test4|
|:--|:--|
|max 22	|max 30 |

**Comment by Spacechild1** [comment]
@Lucarda this doesn't really add up.

In you initial test you had 2% vs 20% in task manager (which basically matches my experience). According to the Pd load meter in your last test both Pds have a similar *high* CPU load.

When I test with Pd 0.51-4, the Pd load meter shows 0 most of the time, but with Pd 0.52-test4 it shows up to 30.

**Comment by Lucarda** [comment]
I will rebuild test4 from scratch on Windows.

On Linux using Pd load meter I got


|Pd 0.51-4|	Pd 0.52-test4|
|:--|:--|
|max 4	|max 30 |

**Comment by Lucarda** [comment]
@Spacechild1 all the values above were double checked on Win10. (i had just rebuilt "test4" via direct download just to be clear that i'm not wrongly using git)

And Pd load meter gives me ~22 on 0.51-4

**Comment by Spacechild1** [comment]
> And Pd load meter gives me ~22 on 0.51-4

That is strange. At least you can still see the regression on Linux.

**Comment by Ant1r** [comment]
sorry, same than before:
CPU forced to 1GHz (Intel® Core™2 Duo CPU T5850)


| version      | moving the mouse | click-draging a numbox |
|        ------   |         -----                 |              ----                    |
| 0.52-test4 |          40                  |         4                            |
| 0.51-4       |           5                   |          5                           |

**Comment by Lucarda** [comment]
@Spacechild1 

There are things I can't explain:

1. Why did i said:

    - "Pd load meter gives me ~22 on 0.51-4" (2 days ago)
  
    - ~2% on 0.51-4 (also 2 days ago)
  
2. Why I don't see the regression on Win10 but I see it on Linux.

</br>
</br>
</br>
</br>




Here's a [video](https://nc.nubegris.com.ar/index.php/s/XNXbnJLxcjF3Y7D) (an `.mp4` that won't play on Firefox but it can be downloaded) of what I get on my win10 machine. 


Video explanation:

3 different Pd Windows 64bit versions (all from Miller's page) running the same patch while monitoring cpu usage via "Pd load meter" and "Windows task manager"

The regression (larger cpu consumption) is only seen on "0.52-test2" when there where "glitching scrollbars". 

I'm still scratching my head.

**Comment by millerpuckette** [comment]
I think this is happening because "gatoms" are now full-fledged "text" objects and so "gobj_getrect" now has to go through the "rtext" formatting process to find the size - before, it was just a matter of looking up box wodth, height, and font size.
If this is truly annoying I can special-case it to save CPU cycles.  OTOH a better, longer-term solution would be to cache hit-relevant data for all visible objects when a cancas becomes visible.

**Comment by Spacechild1** [comment]
>  think this is happening because "gatoms" are now full-fledged "text" objects and so "gobj_getrect" now has to go through the "rtext" formatting process to find the size

I can confirm that I get similar bad performance when using object boxes instead of gatoms! Now I understand what's going on.

>  OTOH a better, longer-term solution would be to cache hit-relevant data for all visible objects when a cancas becomes visible.

+1
